### PR TITLE
Add Function declarations and improve Module declarations

### PR DESCRIPTION
**Pull request by mppf**
* Adds Function, FunctionDecl, Formal, FormalDecl uAST classes
* Renames several enum members to be shorter when the interpretation is clear and otherwise be better worded
  * `VISIBILITY_DEFAULT` - > `DEFAULT_VISIBILITY`
  * `VISIBILITY_PUBLIC` -> `PUBLIC`
  * `VISIBILITY_PRIVATE` -> `PRIVATE`
  * One implication of this strategy is that `Function::CONST_REF` and `Formal::CONST_REF` have the same name (at least if the scoping is not considered). This PR adds a test to also check that they have the same value, in order to reduce confusion in the future, since they refer to the same concept. However, note that I am not expecting code to rely on these having the same value.
* Renames several enum types and related functions named `Tag` (nested in classes) to `Kind` because `Tag`/ `tag()` refers to the ASTNode tag (indicating which class it is)
* Updated ParserContext to track a declaration start location and to tidy up the code setting visibility or variable kind.
* Improved parsing for several variants of identifiers so e.g. `int` will now parse.

### CODE REVIEW

File: compiler/next/include/chpl/uast/Function.h [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+   Return the number of statements in the function body.
+   */
+  int numStmts() const {
+    return numBodyStmts_;
+  }
+
+  /**
+   Return the i'th statement in the function body.
+   */
+  const Expression* stmt(int i) const {
```
Comment (mppf) [inline compiler/next/include/chpl/uast/Function.h:315]: I'm interested for suggestions for the name of this function. Is `stmt` clear enough? Should it mention `body` ? E.g. `bodyStmt(int i)` ? Or is there something better?

File: compiler/next/include/chpl/uast/Function.h [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+  // TODO: should formals() result in Formal or FormalDecl?
+  //       should getFormal() result in the same type?
+  //       do we need two variants of these?
+
+  /**
+   Return a way to iterate over the formals, including the method
+   receiver, if present, as the first formal.
+   */
+  ASTListIteratorPair<FormalDecl> formals() const {
+    if (numFormals_ == 0) {
```
Comment (lydia-duncan) [inline compiler/next/include/chpl/uast/Function.h:199]: It might be good to use the numFormals() method instead of the underlying value - we'll better avoid code drift if we use the same interface we present to outside the class

File: compiler/next/include/chpl/uast/Function.h [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    } else {
+      return nullptr;
+    }
+  }
+
+  /**
+   Return a way to iterate over the lifetime clauses.
+   */
+  ASTListIteratorPair<Expression> lifetimeClauses() const {
+    if (numLifetimeParts_ == 0) {
```
Comment (lydia-duncan) [inline compiler/next/include/chpl/uast/Function.h:267]: Ditto here and in lifetimeClause(), but with numLifetimeClauses() instead

File: compiler/next/include/chpl/uast/FunctionDecl.h [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  FunctionDecl(owned<Function> function)
+    : Decl(asttags::FunctionDecl, std::move(function)) {
+  }
+  bool contentsMatchInner(const ASTNode* other) const override;
+  void markUniqueStringsInner(Context* context) const override;
+
+ public:
+  ~FunctionDecl() override = default;
+
+  // if isMethod is true, the first formal is the 'this' formal
```
Comment (lydia-duncan) [inline compiler/next/include/chpl/uast/FunctionDecl.h:58]: I don't see isMethod in the argument list for this builder (I'm assuming because methods aren't supported yet).  Just pointing that out

File: compiler/next/include/chpl/uast/Sym.h [hunk 5]
```diff
@@ -34,9 +34,9 @@ class Sym : public ASTNode {
 
  public:
   enum Visibility {
-    VISIBILITY_DEFAULT,
-    VISIBILITY_PUBLIC,
-    VISIBILITY_PRIVATE,
+    DEFAULT_VISIBILITY,
```
Comment (lydia-duncan) [inline compiler/next/include/chpl/uast/Sym.h:37]: I think you could shorten this to `DEFAULT_VIS` if you really wanted to

**Review by lydia-duncan — APPROVED** [review]

File: compiler/next/include/chpl/uast/FunctionDecl.h (hunk 4 above)
Comment (mppf) [reply compiler/next/include/chpl/uast/FunctionDecl.h:58]: Actually it's because `owned<FormalDecl> receiver,` indicates if it is a method. The comment is out of date & I'll remove it.

I don't have much testing for methods yet, so they're not fully formed yet, it's true.

File: compiler/next/include/chpl/uast/Function.h (hunk 1 above)
Comment (mppf) [reply compiler/next/include/chpl/uast/Function.h:351]: See https://github.com/Cray/chapel-private/issues/1978
